### PR TITLE
Interface consul etcd

### PR DESCRIPTION
**Pull request by sharadregoti**

### CODE REVIEW

File: utils/syncman/operations.go [hunk 1]
```diff
@@ -104,23 +102,12 @@ func (s *Manager) CreateProjectConfig(project *config.Project) (error, int) {
...
-	opts = opts.WithContext(ctx)
-
-	data, _ := json.Marshal(project)
-
-	_, err := s.consulClient.KV().Put(&api.KVPair{
-		Key:   fmt.Sprintf("sc/projects/%s/%s", s.clusterID, project.ID),
-		Value: data,
-	}, opts)
-	return err, http.StatusInternalServerError
+	s.cb(s.projectConfig)
```
Comment (YourTechBud) [inline utils/syncman/operations.go]: This line isn't required. we are already calling it above

File: utils/syncman/syncman.go [hunk 2]
```diff
@@ -19,43 +16,47 @@ type Manager struct {
...
+	clusterID     string
+	advertiseAddr string
+	port          int
 
 	// Configuration for clustering
-	isConsulEnabled bool
-	consulClient    *api.Client
-	services        []*api.ServiceEntry
+	storeType string
+	storeAddr string
```
Comment (YourTechBud) [inline utils/syncman/syncman.go]: We don't need to store this

File: utils/syncman/store_etcd.go [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+	adminUser  string
+	adminPass  string
+	caCert     string
+	publicKey  string
+	privateKey string
+	isSSL      bool
+}
+
+func NewETCDStore(nodeID, clusterID, advertiseAddr string) (*ETCDStore, error) {
+	etcdConf, err := loadConfig()
```
[the remark this answers is not in the document]
Comment (YourTechBud) [reply utils/syncman/store_etcd.go]: Checkout consuls DefaultConfig for reference

File: utils/syncman/store_consul.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+	if err != nil {
+		log.Fatal("Could not create a new session with consul:", err)
+	}
+
+	data := []byte(s.advertiseAddr)
+	if _, _, err := s.consulClient.KV().Acquire(&api.KVPair{Session: id, Key: fmt.Sprintf("sc/instances/%s/%s", s.clusterID, s.nodeID), Value: data}, opts); err != nil {
+		log.Fatal("Could not register space cloud with consul:", err)
+	}
+
+	ticker := time.NewTicker(3 * time.Second)
```
Comment (YourTechBud) [inline utils/syncman/store_consul.go]:
```suggestion
	ticker := time.NewTicker(4 * time.Second)
```

File: utils/syncman/store_etcd.go [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+	}
+
+	if _, err := s.kv.Put(ctx, fmt.Sprintf("sc/instances/%s/%s", s.clusterID, s.nodeID), s.advertiseAddr, clientv3.WithLease(lease.ID)); err != nil {
+		log.Fatal("Could not register space cloud with etcd:", err)
+	}
+
+	ticker := time.NewTicker(3 * time.Second)
+
+	go func() {
+		for {
```
Comment (YourTechBud) [inline utils/syncman/store_etcd.go]: You do not need a for loop here

File: utils/syncman/store_etcd.go [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+		itemsMeta[id] = &trackedItemMeta{createRevision: kv.CreateRevision, modRevision: kv.ModRevision, project: project}
+	}
+
+	ch := s.etcdClient.Watch(context.Background(), fmt.Sprintf("sc/projects/%s", s.clusterID), clientv3.WithPrefix())
+
+	go func() {
+		for watchResponse := range ch {
+
+			for _, event := range watchResponse.Events {
+				// Return if watch response contains an error
```
Comment (YourTechBud) [inline utils/syncman/store_etcd.go]: Comment is misleading

File: utils/syncman/store_etcd.go [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+					// Remove if incoming mod revision is greater
+					if event.Kv.ModRevision > meta.modRevision {
+						// AddStateless node if doesn't already exists
+						meta.createRevision = event.Kv.CreateRevision
+						meta.modRevision = event.Kv.ModRevision
+						delete(itemsMeta, id)
+					}
+				}
+			}
+			var arrProjects []*config.Project
```
Comment (YourTechBud) [inline utils/syncman/store_etcd.go]: Make this a function and only call.it when there is a change in config

File: utils/syncman/store_etcd.go [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+	}
+
+	for _, kv := range res.Kvs {
+		// Get the id of the item
+		id := strings.Split(string(kv.Key), "/")[idxID]
+		service := &service{id: id, addr: string(kv.Value)}
+		// Store the item
+		itemsMeta[id] = &trackedItemMeta{createRevision: kv.CreateRevision, modRevision: kv.ModRevision, service: service}
+	}
+
```
Comment (YourTechBud) [inline utils/syncman/store_etcd.go]: Call the cb here as well

File: utils/syncman/store_etcd.go [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+						// AddStateless node if doesn't already exists
+						meta.createRevision = event.Kv.CreateRevision
+						meta.modRevision = event.Kv.ModRevision
+						delete(itemsMeta, id)
+					}
+				}
+			}
+
+			// Sort and store
+			var services scServices
```
Comment (YourTechBud) [inline utils/syncman/store_etcd.go]: Put this in a function and only call it when things change

File: main.go [hunk 10]
```diff
@@ -27,18 +27,6 @@ var essentialFlags = []cli.Flag{
 		Usage:  "Load space cloud config from `FILE`",
 		EnvVar: "CONFIG",
 	},
-	cli.StringFlag{
-		Name:   "ssl-cert",
```
[the remark this answers is not in the document]
Comment (sharadregoti) [reply main.go]: ok

File: main.go [hunk 11]
```diff
@@ -54,34 +42,22 @@ var essentialFlags = []cli.Flag{
 		Usage:  "Enable profiler endpoints for profiling",
 		EnvVar: "PROFILER",
 	},
-	cli.StringFlag{
```
[the remark this answers is not in the document]
Comment (sharadregoti) [reply main.go]: ok

File: main.go [hunk 12]
```diff
@@ -162,17 +138,13 @@ func actionRun(c *cli.Context) error {
 	removeProjectScope := c.Bool("remove-project-scope")
 
 	// Load flags related to ssl
-	sslCert := c.String("ssl-cert")
```
[the remark this answers is not in the document]
Comment (sharadregoti) [reply main.go]: ok

File: main.go [hunk 13]
```diff
@@ -220,12 +181,6 @@ func actionRun(c *cli.Context) error {
 	// Initialise the routes
 	s.InitRoutes(profiler, staticPath)
 
-	// Set the ssl config
-	if sslCert != "none" && sslKey != "none" {
```
[the remark this answers is not in the document]
Comment (sharadregoti) [reply main.go]: ok

File: utils/syncman/store_etcd.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sharadregoti) [reply utils/syncman/store_etcd.go]: ok

File: utils/syncman/store_etcd.go [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+				RootCAs:      caCertPool,
+				Certificates: []tls.Certificate{cert},
+			},
+		})
+		if err != nil {
+			return nil, fmt.Errorf("error not able initilize etcd client, %v", err)
+		}
+	} else {
+		client, err = clientv3.New(clientv3.Config{
+			Endpoints: strings.Split(etcdConf.endpoints, ","),
```
[the remark this answers is not in the document]
Comment (sharadregoti) [reply utils/syncman/store_etcd.go]: ok

File: utils/syncman/store_etcd.go [hunk 15]
```diff
@@ -0,0 +1,326 @@
...
+	etcdConf, err := loadConfig()
+	if err != nil {
+		return &ETCDStore{}, fmt.Errorf("error loading etcd config from environment %v", err)
+	}
+
+	var client *clientv3.Client
+	if etcdConf.isSSL {
+		caCert, err := ioutil.ReadFile(etcdConf.caCert)
+		if err != nil {
+			fmt.Errorf("error reading CA cert from provided path,%v", err)
```
[the remark this answers is not in the document]
Comment (sharadregoti) [reply utils/syncman/store_etcd.go]: cannot do that as now we are not returning a variable of pointer type

File: utils/syncman/store_etcd.go [hunk 16]
```diff
@@ -0,0 +1,326 @@
...
+	caCert     string
+	publicKey  string
+	privateKey string
+	isSSL      bool
+}
+
+func NewETCDStore(nodeID, clusterID, advertiseAddr string) (*ETCDStore, error) {
+	etcdConf, err := loadConfig()
+	if err != nil {
+		return &ETCDStore{}, fmt.Errorf("error loading etcd config from environment %v", err)
```
[the remark this answers is not in the document]
Comment (sharadregoti) [reply utils/syncman/store_etcd.go]: cannot do that as now we are not returning a variable of pointer type

File: utils/syncman/store_etcd.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (sharadregoti) [reply utils/syncman/store_etcd.go]: ignore the above

File: utils/syncman/store_etcd.go [hunk 17]
```diff
@@ -0,0 +1,325 @@
...
+	}
+
+	if _, err := s.kv.Put(ctx, fmt.Sprintf("sc/instances/%s/%s", s.clusterID, s.nodeID), s.advertiseAddr, clientv3.WithLease(lease.ID)); err != nil {
+		log.Fatal("Could not register space cloud with etcd:", err)
+	}
+
+	ch, err := s.etcdClient.KeepAlive(context.Background(), lease.ID)
+	if err != nil {
+		log.Println("Could not renew consul session:", err)
+		// register again
```
Comment (YourTechBud) [inline utils/syncman/store_etcd.go]:
```suggestion
```

**Pull request merged**